src/ssh/allowed_signer.rs

```diff
@@ -0,0 +1,289 @@
+        let (mut head, mut rest) = s.split_once(char::is_whitespace)
+            .ok_or(Error::InvalidAllowedSigner("missing key data".to_string()))?;
+
+        let principals: Vec<&str> = head.split(',').collect();
+        if principals.iter().any(|p| p.is_empty()) {
+            return Err(Error::InvalidAllowedSigner("principal cannot be empty".to_string()));
+        }
+        let principals = principals.iter().map(|s| s.to_string()).collect();
+
+        let mut cert_authority = false;
+        let mut namespaces = None;
+        let mut valid_after = None;
+        let mut valid_before = None;
+
+        // We need to trim here and below since split_once(char::is_whitespace) treats
+        // consecutive whitespaces as separate delimiters
+        rest = rest.trim_start();
+        (head, rest) = rest.split_once(char::is_whitespace)
+                .ok_or(Error::InvalidAllowedSigner("missing key data".to_string()))?;
+        loop {
+            // Check if this is a valid option
+            let (option_key, option_value) = match head.split_once('=') {
+                Some(v) => v,
+                None => (head, ""),
+            };
+            match option_key.to_lowercase().as_str() {
+                "cert-authority" => cert_authority = true,
+                "namespaces" => {
+                    if namespaces.is_some() {
+                        return Err(Error::InvalidAllowedSigner("multiple \"namespaces\" clauses".to_string()));
+                    }
+
+                    let (namespaces_value, rest_) = parse_namespaces(option_value, &mut rest)?;
+                    rest = rest_;
+                    let namespaces_value: Vec<&str> = namespaces_value.split(',')
+                        .filter(|e| !e.is_empty())
+                        .collect();
+                    namespaces = Some(
+                        namespaces_value.iter()
+                            .map(|s| s.to_string())
+                            .collect()
+                    );
+                },
+                "valid-after" => {
+                    if valid_after.is_some() {
+                        return Err(Error::InvalidAllowedSigner("multiple \"valid-after\" clauses".to_string()));
+                    }
+                    valid_after = Some(parse_timestamp(option_value)
+                        .map_err(|_| Error::InvalidAllowedSigner("invalid \"valid-after\" time".to_string()))?);
+                },
+                "valid-before" => {
+                    if valid_before.is_some() {
+                        return Err(Error::InvalidAllowedSigner("multiple \"valid-before\" clauses".to_string()));
+                    }
+                    valid_before = Some(parse_timestamp(option_value)
+                        .map_err(|_| Error::InvalidAllowedSigner("invalid \"valid-before\" time".to_string()))?);
+                },
+                // If option_key does not match any valid option, we test if it's the key data
```

Comment:
I think it's better if you pull off the key data first then work on the options. That way if there are new option types in the future it doesn't have to break the parsing (we can choose to fail parsing with unknown options or allow)

Reply:
Comments at the end of the line kinda complicates this... It's not easy to tell where the end starts without test parsing the key from the end to the front. I can't just slice off the last # either since that might be in the namespace.

Reply:
Fair enough